File: hashmap/383_ransom_note-grind75-solved/solution.ts
/*
 * 383. Ransom Note
 * https://leetcode.com/problems/ransom-note/
 *
 * Given two strings ransomNote and magazine, return true if ransomNote can be
 * constructed by using the letters from magazine and false otherwise.
 * Each letter in magazine can only be used once in ransomNote.
 *
 * Example 3: ransomNote="aa", magazine="aab" → true
 * Constraints: 1 <= lengths <= 10^5, lowercase letters only.
 */

function canConstruct(ransomNote: string, magazine: string): boolean {
    const magaMap = new Map<string, number>()
    for (const c of magazine) {
        magaMap.set(c, (magaMap.get(c) || 0) + 1)
    }
    for (const c of ransomNote) {
        if (!magaMap.has(c)) {
            return false;
        } else {
            if (magaMap.get(c)! > 1) {
                magaMap.set(c, magaMap.get(c)! - 1)
            } else {
                magaMap.delete(c)
            }
        }
    }
    return true
};

File: hashmap/383_ransom_note-grind75-solved/solution.py
"""
383. Ransom Note
https://leetcode.com/problems/ransom-note/

Return true if ransomNote can be constructed using letters from magazine (each letter used once).

Example 1: ransomNote="a", magazine="b" → false
Example 2: ransomNote="aa", magazine="ab" → false
Example 3: ransomNote="aa", magazine="aab" → true

Constraints:
    1 <= ransomNote.length, magazine.length <= 10^5
    Both consist of lowercase English letters.
"""


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        maga_map = {}
        for s in magazine:
            maga_map[s] = maga_map.get(s, 0) + 1  # get 代替 if/else

        for c in ransomNote:
            if c not in maga_map:
                return False
            else:
                if maga_map[c] > 1:
                    maga_map[c] -= 1   # -= 代替 = x - 1
                else:
                    del maga_map[c]
